### PR TITLE
fix(eager): compile d2d copy and overwrite statically per shape

### LINKED ISSUE #3826: Device-to-device slice write copies the source's base extent, not the view's size
### What happened?

`dst[a:b] = src` writes the wrong number of rows when `src` is a prefix view of a longer tensor. It copies the view's base extent instead of the view's size, so it overwrites rows past the end of the destination slice.

Eager only. The compiled lowering of the same write is correct.

It also only happens after an earlier prefix-view slice write at a *different* view length has run in the same process. The first such write in a process is always correct.

### What did you expect to happen?

`dst[32:64] = src` should write 32 rows and leave the rest alone, as it does on CPU.

### How can we reproduce it?

fp16, AIU 1.0:

```python
import torch
import torch_spyre  # noqa: F401


def prefix_view_write(mode, qlen):
    """dst[32:32+qlen] = <first qlen rows of a 64-row tensor>.

    Returns (rows changed outside the window, window is correct).
    """

    def fn(dst, src):
        dst[32 : 32 + qlen] = src

    if mode == "compile":
        fn = torch.compile(fn, dynamic=False, backend="inductor")

    src_cpu = torch.randn(64, 32, 128, dtype=torch.float16)
    dst = torch.zeros(96, 32, 128, dtype=torch.float16).to("spyre")
    fn(dst, src_cpu.to("spyre")[:qlen])

    got = dst.cpu()
    outside = torch.ones(96, dtype=torch.bool)
    outside[32 : 32 + qlen] = False
    spill = int((got[outside].abs().amax(dim=(1, 2)) > 1e-3).sum())
    window_ok = torch.allclose(got[32 : 32 + qlen], src_cpu[:qlen], atol=1e-2)
    return spill, window_ok


for mode in ("eager", "compile"):
    print(mode, "warm-up:", prefix_view_write(mode, 16), "under test:", prefix_view_write(mode, 32))
```

Output:

```
eager warm-up: (0, True) under test: (32, True)
compile warm-up: (0, True) under test: (0, True)
```

In eager the second call asks for 32 rows but writes 64, so `dst[64:96]` receives `src`'s rows 32..63. The window `dst[32:64]` stays correct in every case, so only the extra rows are wrong.

Both calls are needed. The second one on its own is correct, in either mode.

### Anything else we need to know?

| source | eager | compile |
|---|---|---|
| exact-size tensor | ok | ok |
| prefix view, first write in the process | ok | ok |
| prefix view, after a prefix-view write at a different length | writes the base extent | ok |
| prefix view + `.contiguous()` | writes the base extent | ok |
| prefix view + `.clone()` | ok | ok |

- The write always ends at `offset + base_extent`.
- `.contiguous()` changes nothing: a prefix slice is already contiguous and keeps the same storage. `.clone()` shrinks the extent and fixes it.
- What triggers it is a change in the source view length. Changing the destination size or the write offset does not.
- Where the overrun runs past the end of the destination tensor, the write leaves the tensor entirely.
- CPU is fine. Same behaviour for `dst[a:b] = src` and `narrow().copy_()`.
- No `FallbackWarning` in either mode, so the compiled case is genuinely correct on device rather than quietly running on CPU.
- The eager path goes through `spyre__copy_from` -> `torch.ops.spyre.copy_from_d2d`, which forwards storage offsets but nothing about the views' sizes.

Found in spyre-inference paged attention: the write-back sliced padding rows off the result, and the over-long write corrupted other sequences in the batch and produced NaN. The write-back is eager, so it is exposed. Worked around with `.clone()` in torch-spyre/spyre-inference#546.

### PR DESCRIPTION
**Pull request by jturney**
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Fixes the eager d2d slice-write corruption in #3826: `dst[a:b] = src` from a prefix view wrote a **frozen extent** instead of the view's size once the process had performed such a write at a *different* view length — silently clobbering memory past the destination window.

**Root cause.** `spyre::copy_from_d2d` is compiled through `compile_once`, which called `torch.compile(op)` without `dynamic=False`. Dynamo's auto-dynamic promotes a **size** to a symbol after repeated calls at distinct values — exactly as it does for ints, which this op already fights off with `specialize_int` — and the Spyre lowering then silently bakes one concrete extent into the SDSC while dynamo reuses the "dynamic" graph for every later size. Measured on a 5-length ladder: after auto-dynamic kicks in, the reused graph writes its trace-time extent regardless of the runtime length (too few rows at 48, an 8-row overrun at 24; on @tdoublep's stack the baked value was the base's 64 rows — same mechanism, different frozen constant). This explains every row of the issue's excellent diagnosis table, including why `.clone()` fixes it, `.contiguous()` doesn't, and the first write in a process is always correct.

**Fix.** `dynamic=False`, so every shape gets its own static trace — the pattern every other `compile_once` site in the tree already uses, in the one-binary-per-variant regime `cache_size_limit=1024` exists for. `spyre::overwrite` (the scatter-write primitive) had the identical exposure — same `specialize_int` battle scars, same missing kwarg — and gets the identical fix.

#### Which issue(s) this PR is related to:

Fixes #3826

#### Special notes for your reviewer:

- The regression test is a **multi-length ladder on purpose**: the first trace is always correct and the corruption only appears once auto-dynamic would have kicked in, so a single-length test structurally cannot regress this. The test asserts both containment (no writes outside the window) and content.
- Verified: the issue's reproducer verbatim, a 5-length ladder (16/32/48/24/40 — all clean, `unique_graphs: 5`), and `test_overwrite.py` + `test_offset_view_eager_ops.py` (28 passed).
- **Deliberately not fixed here:** user-facing `torch.compile(dynamic=True)` on the same write corrupts from the *second* call, because the backend accepts a graph with a symbolic extent and silently specializes it. That is a separate, pre-existing gap in how the backend handles symbolic dims (it should refuse ones it cannot honor rather than freeze them), out of scope for this fix — this PR closes the eager door #3826 walked through, not that one.
- Based on `3e23d180` rather than current tip: `main` @ `ab6eb2f7` currently requires AIUPTI runtime headers (`AIUPTI_RUNTIME_TRACE_CBID_PARSE_RESPONSE`, #3591) newer than our pod's installed toolchain, so local device verification was run on the newest buildable base. No conflict with the intervening commits (`customops.py` auto-merges cleanly over #3811).

#### Does this PR introduce a user-facing change?

Eager d2d slice writes from views are now correct at every view length. Compile-time cost: one compiled copy program per distinct shape instead of a shared dynamic one — the regime the existing `cache_size_limit` bump was designed for.

#### Additional note:

Credit to @tdoublep for the diagnosis table in #3826, which narrowed this to a state-dependence signature before any code was read.

### CODE REVIEW

**Review by dgrove-oss — APPROVED** [review]
Summary

Fixes #3826: spyre::copy_from_d2d and spyre::overwrite were compiled via compile_once without dynamic=False, so dynamo's auto-dynamic promotion silently froze one concrete size into the SDSC after repeated calls at distinct sizes, causing slice writes from prefix views to write a stale extent — a silent memory-corruption bug in attention write-back. The fix adds dynamic=False to both compile_once call sites, matching the one other pre-existing site in eager.py:285 that already had it.

Review

BLOCKERs

None.

SUGGESTIONs

None — the fix is minimal and precisely targeted.

QUESTIONs

None.

Verification performed

- Root-cause mechanism checks out. Verified compile_once(op, **compile_kwargs) (torch_spyre/ops/eager.py:29) forwards kwargs straight to torch.compile, so dynamic=False is a real, correctly-plumbed fix, not a guess.
- "Every other site already does this" claim confirmed. Grepped all compile_once( call sites in the tree — only 3 exist total; the pre-existing one (eager.py:285) already passes dynamic=False, and the two being fixed here were the only outliers.
- specialize_int vs. dynamic=False are orthogonal, both needed. specialize_int=True (already present, from PR #2084) guards scalar int args (offsets/dims); it doesn't touch dynamo's auto-dynamic promotion of tensor size dims — the actual gap. The fix is additive and doesn't remove or fight the existing specialize_int patch.
- New test is genuinely novel coverage, not overlapping existing tests. test_overwrite.py's loops vary offsets/slots (already covered by specialize_int), not input tensor size — confirmed no existing test exercises the multi-size-ladder scenario this PR fixes.
- Test correctly targets the fixed code path. dst[32:32+qlen] = src.to("spyre")[:qlen] on a view goes through the offset-materialization path documented at the top of test_offset_view_eager_ops.py, which routes into copy_from_d2d.
- Pre-commit (ruff, mypy, import-regex enforcement) passes on both changed files; license header on the new test file is correct (2026).
- DCO sign-off present.
- Full CI is green, including the new/existing Test Offset View Eager Ops and Test Overwrite suites.

Verdict

APPROVE
